src/AST/Translator.Entity.dfy

```diff
@@ -0,0 +1,298 @@
+    var ty :- Expr.TranslateType(st.Rhs);
+    var constraint :-Expr.TranslateExpression(st.Constraint);
+    var wit :-Expr.TranslateOptionalExpression(st.Witness);
+    var ei :- TranslateTopLevelEntityInfo(st);
+    Success([E.Entity.Type(ei, E.Type.SubsetType(E.SubsetType.SubsetType(x, ty, constraint, wit)))])
+  }
+
+  function TranslateTypeParameter(ts: C.TypeParameter): (e: TranslationResult<seq<E.Entity>>)
+    reads *
+  {
+    // TODO: handle variance, etc
+    var ei :- TranslateTopLevelEntityInfo(ts);
+    Success([E.Entity.Type(ei, E.Type.TypeParameter(E.TypeParameter.TypeParameter()))])
+  }
+
+  function TranslateOpaqueTypeDecl(ot: C.OpaqueTypeDecl): (e: TranslationResult<E.Type>)
+    reads *
+  {
+    Success(E.Type.AbstractType(E.AbstractType.AbstractType()))
+  }
+
+  function TranslateNewtypeDecl(nt: C.NewtypeDecl): (e: TranslationResult<E.Type>)
+    reads *
+  {
+    var x := TypeConv.AsString(nt.Var.Name);
+    var ty :- Expr.TranslateType(nt.BaseType);
+    var constraint :- Expr.TranslateOptionalExpression(nt.Constraint);
+    var wit :-Expr.TranslateOptionalExpression(nt.Witness);
+    Success(E.Type.NewType(E.NewType.NewType(x, ty, constraint, wit)))
+  }
+
+  function TranslateDatatypeDecl(dt: C.DatatypeDecl): (e: TranslationResult<E.Type>)
+    reads *
+  {
+    Success(E.Type.DataType(E.DataType.DataType()))
+  }
+
+  function TranslateTraitDecl(t: C.TraitDecl): (e: TranslationResult<E.Type>)
+    reads *
+  {
+    var parentTraits :- Seq.MapResult(ListUtils.ToSeq(t.ParentTraits), (t: C.Type) reads * =>
+      TranslateName(t.AsTraitType.FullName));
+    Success(E.Type.TraitType(E.TraitType.TraitType(parentTraits)))
+  }
+
+  function TranslateClassDecl(c: C.ClassDecl): (e: TranslationResult<E.Type>)
+    reads *
+  {
+    var parentTraits :- Seq.MapResult(ListUtils.ToSeq(c.ParentTraits), (t: C.Type) reads * =>
+      TranslateName(t.AsTraitType.FullName));
+    Success(E.Type.ClassType(E.ClassType.ClassType(parentTraits)))
+  }
+
+  function TranslateTopLevelEntityInfo(tl: C.TopLevelDecl): (e: TranslationResult<E.EntityInfo>)
+    reads *
+  {
+    var name :- TranslateName(tl.FullName);
+    var attrs :- TranslateAttributes(tl.Attributes);
+    var loc := TranslateLocation(tl.tok);
+    Success(E.EntityInfo(name, location := loc, attrs := attrs, members := []))
+  }
+
+  function TranslateTopLevelEntityInfoMembers(tl: C.TopLevelDeclWithMembers): (e: TranslationResult<(seq<E.Entity>, E.EntityInfo)>)
+    reads *
+  {
+    var name :- TranslateName(tl.FullName);
+    var attrs :- TranslateAttributes(tl.Attributes);
+    var loc := TranslateLocation(tl.tok);
+    var memberDecls := ListUtils.ToSeq(tl.Members);
+    var members :- Seq.MapResult(memberDecls, d reads * => TranslateMemberDecl(d));
+    var memberNames := Seq.Map((m: E.Entity) => m.ei.name, members);
+    :- Need(forall nm <- memberNames :: nm.ChildOf(name), Invalid("Malformed member name in " + name.ToString()));
```

Comment:
Would it work to run the validator after the fact instead of having `Need` here?

Reply:
Turns out that that's the wrong set of member names. We need to filter out the ones that are direct children, and when we do that we immediately get the necessary fact to construct a valid `EntityInfo`. :tada: